### PR TITLE
Add firebase

### PR DESCRIPTION
**Pull request by vs2961**
## Tracking Info

Resolves parts of #15

## Changes

Added firebase to frontend and verify tokens in backend.

## Testing

Tested with logging in with abc@testing.com and password abc123, then verifying the account's info in compass.

## Confirmation of Change

See the testing section.

### CODE REVIEW

**Comment by vs2961** [comment]
Waiting on @parth4apple to change the .env firebase variable to be named NEXT_PUBLIC_FIREBASE_CONFIG instead, because for some reason you can't read env variables that don't start with NEXT_PUBLIC.,

File: frontend/src/pages/login.tsx [hunk 1]
```diff
@@ -1,19 +1,69 @@
...
+        return userCredential.user.getIdToken();
+      })
+      .then((token) => {
+        return token;
+      });
+  };
+
+  const sendTokenToBackend = async (token: string) => {
+    try {
+      const response = await GET(`http://localhost:4000/api/user/`, {
```
Comment (adhi0331) [inline frontend/src/pages/login.tsx]: Might want to change the url to not be localhost

File: frontend/src/pages/login.tsx [hunk 2]
```diff
@@ -1,19 +1,69 @@
...
   const onSubmit: SubmitHandler<FieldValues> = (data) => {
     console.log(data);
+    login(data.email as string, data.password as string)
+      .then((token: string) => {
+        void sendTokenToBackend(token);
+      })
+      .catch((_) => {
+        setFirebaseError("Invalid login. Please check your username and password.");
+      });
+    router.push("/profile");
```
Comment (adhi0331) [inline frontend/src/pages/login.tsx]: let's route to the home page instead of profile

File: frontend/src/pages/login.tsx [hunk 3]
```diff
@@ -1,19 +1,69 @@
...
+  } = useForm();
   const _setValue = setValue;
 
+  const [firebaseError, setFirebaseError] = useState("");
+
+  const { auth } = initFirebase();
+  const router = useRouter();
+
+  const login = async (email: string, password: string) => {
+    return await signInWithEmailAndPassword(auth, email, password)
```
Comment (adhi0331) [inline frontend/src/pages/login.tsx:30]: Screenshot 2024-04-03 at 9 40 34 AM

When I login in on google chrome I get this error message from google. Not sure what is causing this but it is something to look into.

**Review by adhi0331 — CHANGES_REQUESTED** [review]
Looks really good! Left a few comments, once those are addressed should be good to merge.

File: frontend/src/pages/login.tsx (hunk 3 above)
Comment (vs2961) [reply frontend/src/pages/login.tsx:30]: it's just that the sample password we used, abc123, is a pretty insecure password and google doesn't like that

File: frontend/src/pages/login.tsx (hunk 1 above)
Comment (vs2961) [reply frontend/src/pages/login.tsx]: Do you want me to put the backend URL as a env variable?

Comment (adhi0331) [reply frontend/src/pages/login.tsx:42]: If you look at this [line](https://github.com/TritonSE/PIA-Program-Manager/blob/main/frontend/src/api/requests.ts#L10) on main we already have the different base URL configurations setup. I think you should be able to call the route by doing `GET('user/', ...)`. You can look at how the [student page ](https://github.com/TritonSE/PIA-Program-Manager/blob/main/frontend/src/api/students.ts#L47) works for reference.

Comment (adhi0331) [reply frontend/src/pages/login.tsx]: Realized that there was an error in main because the base url was incorrect. Fixed it tho

Comment (adhi0331) [reply frontend/src/pages/login.tsx]: I think it was supposed to be an env variable. Not sure if it got deleted or never modified on the google drive.

Comment (adhi0331) [reply frontend/src/pages/login.tsx]: Yeah that would be good. I tried modifying the URL but when I did it looks like this function throws some error. Essentially after the logging in the home page student table doesn't load. However, if I were to access the home page without logging in then the student table does load, so I think there's some bug with this function and not the student table. Ideally the request URL should just be something like `/user/` for this function.

**Review by adhi0331 — COMMENTED** [review]
One last thing that needs to be fixed, right now users can access the main pages (home, programs, notifications, etc) without actually logging in. They can just type in the corresponding url and retrieve the data. Obviously we do not want that so we need to ensure that for each page that an eligible user accesses it. If you have the bandwidth we also need to ensure that each page is different based on user roles (for example only admins can access certain pages like notifications, see figma for details).

File: frontend/src/pages/login.tsx [hunk 4]
```diff
@@ -1,19 +1,69 @@
...
   const onSubmit: SubmitHandler<FieldValues> = (data) => {
     console.log(data);
+    login(data.email as string, data.password as string)
+      .then((token: string) => {
+        void sendTokenToBackend(token);
+      })
+      .catch((_) => {
+        setFirebaseError("Invalid login. Please check your username and password.");
+      });
+    router.push("/home");
```
Comment (adhi0331) [inline frontend/src/pages/login.tsx]: When I tried logging in with an invalid username and password it still routed to the home page.

**Review by adhi0331 — CHANGES_REQUESTED** [review]
Noticed a bug when playing around with it a little more.

**Review by adhi0331 — APPROVED** [review]
LGTM!!! 🚀

**Pull request merged**